src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java

```diff
@@ -98,6 +98,48 @@ public void onFailure(Throwable e) {
         });
     }
 
+    public void attemptSyncedFlush(final String[] aliasesOrIndices, IndicesOptions indicesOptions, final ActionListener<IndicesSyncedFlushResult> listener) {
+        final ClusterState state = clusterService.state();
+        final String[] concreteIndices = state.metaData().concreteIndices(indicesOptions, aliasesOrIndices);
+        final Map<String, List<ShardsSyncedFlushResult>> results = ConcurrentCollections.newConcurrentMap();
+        int totalNumberOfShards = 0;
+        int numberOfShards = 0;
+        for (String index : concreteIndices) {
+            final IndexMetaData indexMetaData = state.metaData().index(index);
+            totalNumberOfShards += indexMetaData.totalNumberOfShards();
+            numberOfShards += indexMetaData.getNumberOfShards();
+            results.put(index, Collections.synchronizedList(new ArrayList<ShardsSyncedFlushResult>()));
+
+        }
+        final int finalTotalNumberOfShards = totalNumberOfShards;
+        final CountDown countDown = new CountDown(numberOfShards);
+
```

Comment:
flush request will hang if there are no concrete indices. I think we have to check this case and then call listener.onResponse()